ckan/lib/uploader.py

```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
                 os.remove(filepath)
             except OSError:
                 pass
+
+
+class FkUpload(object):
+    storage: fk.Storage | None = None
+    filename: Optional[str]
+    object_type: Optional[str]
+    old_filename: Optional[str]
+    upload_file: files.Upload | None
+
+    def __init__(self,
+                 object_type: str,
+                 old_filename: Optional[str] = None) -> None:
+        ''' Setup upload by creating a subdirectory of the storage directory
+        of name object_type. old_filename is the name of the file in the url
+        field last time'''
+        self.filename = None
+        self.object_type = object_type
+        self.old_filename = old_filename
+
+        storage_name: str = config[f"ckan.files.default_storages.{object_type}"]
+        try:
+            self.storage = files.get_storage(storage_name)
+        except files.exc.UnknownStorageError:
+            log.warning(
+            "Storage %s is not configured and upload of %s will be ignored",
+                storage_name,
+                object_type,
+            )
+
+    def update_data_dict(self, data_dict: dict[str, Any], url_field: str,
+                         file_field: str, clear_field: str) -> None:
+        ''' Manipulate data from the data_dict.  url_field is the name of the
+        field where the upload is going to be. file_field is name of the key
+        where the FieldStorage is kept (i.e the field where the file data
+        actually is). clear_field is the name of a boolean field which
+        requests the upload to be deleted.  This needs to be called before
+        it reaches any validators'''
+
+        self.url = data_dict.get(url_field, '')
+        self.clear = data_dict.pop(clear_field, None)
+        self.file_field = file_field
+        self.upload_field_storage = data_dict.pop(file_field, None)
+
+        if not self.storage:
+            return
+
+        if isinstance(self.upload_field_storage, ALLOWED_UPLOAD_TYPES):
+            if self.upload_field_storage.filename:
+                self.filename = self.upload_field_storage.filename
+                self.filename = str(datetime.datetime.utcnow()) + self.filename
+                self.filename = munge.munge_filename_legacy(self.filename)
```

Comment:
`munge_filename_legacy` is the function that replaces all non-ascii characters, which is really bad for many users.